realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt

```diff
@@ -0,0 +1,510 @@
+package io.realm
+
+import io.realm.entities.AllTypes
+import io.realm.entities.Dog
+import io.realm.entities.SimpleClass
+import io.realm.kotlin.*
+import io.realm.rule.TestRealmConfigurationFactory
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.util.concurrent.CountDownLatch
+import kotlin.test.*
+
+@ExperimentalCoroutinesApi
+class CoroutineTests {
+
+    @Suppress("MemberVisibilityCanPrivate")
+    @Rule
+    @JvmField
+    val configFactory = TestRealmConfigurationFactory()
+
+    private lateinit var configuration: RealmConfiguration
+    private lateinit var testDispatcher: TestCoroutineDispatcher
+    private lateinit var testScope: TestCoroutineScope
+    private lateinit var realm: Realm
+
+    @Before
+    fun setUp() {
+        testDispatcher = TestCoroutineDispatcher()
+        testScope = TestCoroutineScope(testDispatcher)
+        configuration = configFactory.createConfiguration()
+        realm = Realm.getInstance(configuration)
+    }
+
+    @After
+    fun tearDown() {
+        realm.executeTransaction { it.deleteAll() }
+        realm.close()
+    }
+
+    @Test
+    fun toFlow_emittedOnCollect() {
+        val countDownLatch = CountDownLatch(1)
+
+        // TODO check this out for better testing: https://proandroiddev.com/from-rxjava-to-kotlin-flow-testing-42f1641d8433
+        val context = Dispatchers.Main
+        val scope = CoroutineScope(context)
+
+        scope.launch {
+            val realmInstance = Realm.getInstance(configuration)
+
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        assertEquals(0, flowResults.size)
+                        scope.cancel("Cancelling scope...")
+                    }.onCompletion {
+                        realmInstance.close()
+                        countDownLatch.countDown()
+                    }.launchIn(scope)
```

Comment:
Maybe this one is not the best way to start collecting results from a flow, but `launchIn` is shorthand for `scope.launch { flow.collect() }`. The advantage is that it returns a `Job`, so that may come in handy for other purposes. But yeah, in this case I'm not doing anything with the job, so I might as well use collect instead.

Reply:
There is a difference though, which is `launchIn` is not a coroutine and `collect` is.

Reply:
This actually means that if we call `collect` and the flow isn't cancelled somehow, it will never end and will hog the thread on which it's running, which isn't a problem here, but it is for instance in `toFlow_resultsCancelBeforeCollectActualResults` since we are doing things within the coroutine scope after calling collect.